Content.Shared/Replays/ReplayEvent.cs

```diff
@@ -0,0 +1,231 @@
+﻿using System.Numerics;
+using Content.Server.Replays;
+using Content.Shared.Mobs;
+
+namespace Content.Shared.Replays;
+
+/// <summary>
+/// Represents an event that occured during gameplay, such as a chat message, a player joining, etc.
+/// with the goal of being able to reconstruct a round without actually running the replay (so external tools parsing it, without knowing the specifics of the fork)
+/// </summary>
+[Serializable, DataDefinition]
+public partial class ReplayEvent
+{
+    /// <summary>
+    /// Seconds into the round when this event occured.
+    /// </summary>
+    [DataField]
+    public double? Time;
+
+    [DataField]
+    public ReplayEventSeverity? Severity;
+
+    [DataField]
+    public ReplayEventType? EventType;
+
+    [DataField]
+    public LocationInformation? Location;
+};
+
+[Serializable, DataDefinition]
+public sealed partial class LocationInformation
+{
+    [DataField]
+    public string? NearestBeacon;
+
+    [DataField]
+    public Vector2? Position;
+
+    [DataField]
+    public string? Map;
+}
+
+/// <summary>
+/// A generic player event that can be used for any player-related event.
+/// </summary>
+[Serializable, DataDefinition]
+public sealed partial class GenericPlayerEvent : ReplayEvent
+{
+    /// <summary>
+    /// The player info associated with this event. This who this event is about.
+    /// </summary>
+    /// <remarks>
+    /// Storing the player name etc. every time is a bit wasteful, but since replays are already big as fuck, it's not that big of a deal.
+    /// </remarks>
+    [DataField]
+    public ReplayEventPlayer? Target;
+
+    /// <summary>
+    /// The source of the event. Who was the cause for the Target being affected? Can be null.
+    /// </summary>
+    [DataField]
+    public ReplayEventPlayer? Origin;
+}
+
+/// <summary>
+/// A generic event used for object or non player entites. It does not contain player info, and only strings.
+/// </summary>
+[Serializable, DataDefinition]
+public sealed partial class GenericObjectEvent : ReplayEvent
+{
+    [DataField]
+    public string? Target;
```

Comment:
Not sure how I feel about just logging entity IDs in fields like this, seems short-sighted for the future.